prusti-common/src/vir/optimizations/predicates/delete_unused_predicates.rs

```diff
@@ -0,0 +1,191 @@
+// © 2020, ETH Zurich
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+use crate::vir::{ast::*, cfg, cfg::CfgMethod, CfgBlock};
+use std::collections::BTreeSet;
+
+fn get_used_predicates(methods: &[CfgMethod], functions: &[Function]) -> BTreeSet<String> {
+    let mut collector = UsedPredicateCollector::new();
+    super::walk_methods(methods, &mut collector);
+    super::walk_functions(functions, &mut collector);
+
+    // DeadBorrowToken$ is a used predicate but it does not appear in VIR becaue it is only created when viper code is created from VIR
+    collector
+        .used_predicates
+        .insert("DeadBorrowToken$".to_string());
+    collector.used_predicates
+}
+
+fn get_used_predicates_in_predicates(predicates: &[Predicate]) -> BTreeSet<String> {
+    let mut collector = UsedPredicateCollector::new();
+
+    for pred in predicates {
+        match pred {
+            Predicate::Struct(StructPredicate { body: Some(e), .. }) => {
+                ExprWalker::walk(&mut collector, e)
+            }
+            Predicate::Struct(_) => { /* ignore */ }
```

Comment:
I think it would be better to make the code more explicit.
```suggestion
            Predicate::Struct(StructPredicate { body: None, .. }) => { /* ignore */ }
```